### PR TITLE
add default hash to ndarray

### PR DESCRIPTION
**Pull request by szha**
## Description ##
Define `__hash__` for NDArray for consistent py2 py3 behavior. The included hash function is the default in python2.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] For user-facing API changes, API doc string has been updated.
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Add `__hash__` to `NDArray`

### CODE REVIEW

File: python/mxnet/ndarray/ndarray.py [hunk 1]
```diff
@@ -286,6 +286,10 @@ def __eq__(self, other):
         """x.__eq__(y) <=> x==y <=> mx.nd.equal(x, y) """
         return equal(self, other)
 
+    def __hash__(self):
+        """Default hash function."""
+        return id(self)//16
```
Comment (piiswrong) [inline python/mxnet/ndarray/ndarray.py]: why //16?

Comment (szha) [reply python/mxnet/ndarray/ndarray.py]: This happens to be the default hash function for python objects when `__eq__` is not defined.

Comment (larroy) [reply python/mxnet/ndarray/ndarray.py]: Does this still apply?


You should add a comment to this, it's unclear.

In [1]: import mxnet as mx
^[[A
In [2]: b=mx.nd.array([1,2,3,4])

In [3]: a=mx.nd.array([1,2,3,4,5])

In [4]: hex(id(a))
Out[4]: '0x11d39df60'

In [5]: hex(id(b))
Out[5]: '0x11d383c88'


https://stackoverflow.com/questions/11324271/what-is-the-default-hash-in-python

Comment (szha) [reply python/mxnet/ndarray/ndarray.py]: @larroy why does it matter?

Comment (larroy) [reply python/mxnet/ndarray/ndarray.py]: Is the division by 16 needed? If yes i would comment on why. If not, I would just remove it. It causes confusion to use such a constant without further explanation or a clarifying comment.

**Pull request merged**